### DISCUSSION TITLE
How to use AdbDaemonDirectSocketsDevice and TCPSocket class to create connection?

### DISCUSSION THREAD

**Discussion by wbwxshao** [yume-chan/ya-webadb#613] [General]
Hi all!
I found the two classes here: https://tango-adb.github.io/docs/tango/daemon/tcp/create-connection/
I would like to use these two classes and create wireless adb connection to Android devices. I already have the IP address and port details. When trying to implement this feature, I was not able to directly import AdbDaemonDirectSocketsDevice and TCPSocket class so I figured we need to create the class manually as shown in the doc. But there are errors when copying & pasting the code I saw several errors. For example, the below code 
`const readable = new PushReadableStream((controller) => {
        this.#socket.on('data', async (data) => {
          this.#socket.pause();
          await controller.enqueue(data);
          this.#socket.resume();
        });`
is not assignable when sending into this.#opened.resolve()

`this.#opened.resolve({
        remoteAddress,
        remotePort,
        localAddress: this.#socket.localAddress!,
        localPort: this.#socket.localPort!,
        readable,
        writable: new WritableStream({
          write: async (chunk) => {
            return new Promise<void>((resolve, reject) => {
              if (!this.#socket.write(chunk)) {
                this.#socket.once('drain', resolve);
              } else {
                resolve();
              }
            });
          },
          close: async () => {
            this.#socket.end();
          },
        }),
      });
    });`

Any help will be appreciated!

**Comment by yume-chan** [comment]
I updated and tested the example code in that page. It should work now.

**Reply by wbwxshao** [reply]
Thanks a lot! Now it doesn't have any error.
One more question, is the Node.js code intended for a backend server? I tried to integrate the code for TCPSocket in my Angular frontend but when building got this error:
node:net - Error: Module build failed: UnhandledSchemeError: Reading from "node:net" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "node:" URIs.

**Reply by yume-chan** [reply]
Of course you can't use Node.js built in modules in Web browsers, but they are also not limited to backend servers, they are also available in Electron.

**Reply by wbwxshao** [reply]
Thanks for your answer! Would you mind looking at our other discussion here: https://github.com/yume-chan/ya-webadb/discussions/609#discussioncomment-7523736 
Please let us know if you are interested or not, we will not bother you anymore. Thanks!
